### ISSUE TITLE
Memory leak amCharts 5 and amCharts 4

### ISSUE THREAD

**Issue by deicidemilan** [amcharts/amcharts5#125]
Hi,

Recently I start to implement and test amCharts 5, because I search for good and stable chart lib.
I have issues with Memory leak in all charts on my app.
Then I go to codepen example, and find there are Memory leak too.
Then i go to official amCharts site, and find Memory leak on charts.
This example [Official amCharts demo](https://www.amcharts.com/demos/line-chart-adding-data-every-second)
I found that Memory leak happened when I have animation, zoom and pan actions.
Is this a bug or this is a free version of lib?
Even on amCharts 4 I got Memory leak.

Best regards,
Milan

**Comment by deicidemilan** [comment]
I try version 5.0.15 and same thing, in 10-15 minutes lib eats 2gb of memory and chrome crash tab.
I will demonstrate how I got this, for you guys to try fix this.
Memory leak in my app start immediately as I start app. In your demos on official page i need to zoom and pan few times, and memory start to grow, and GC can't clear in time.
I noticed bug happens only if  I have animation, zoom and pan actions, as i mentioned earlier.

This is a picture of chrome Task Manager Tab is from official site demo.

![This is an image](https://i.ibb.co/SXLrhrm/amcharts.png)

**Comment by Pauan** [comment]
@deicidemilan We did extensive testing and we found no memory leak. Even after 15 minutes, the memory stayed consistent at 130k. In addition, the heap snapshot shows no leaked objects.

![Screenshot 2021-12-08 09:19:52](https://user-images.githubusercontent.com/97238/145178328-bdd87dac-1be9-4ce9-a9fb-249ace523e06.png)

Make sure that you are using version 5.0.15 (delete your `node_modules` folder, and your `yarn.lock` and `package-lock.json` files).

Also, make sure that you are testing this code (*and no other code*):

```js
import * as am5 from "@amcharts/amcharts5";
import * as am5xy from "@amcharts/amcharts5/xy";
import am5themes_Animated from "@amcharts/amcharts5/themes/Animated";

var root = am5.Root.new("chartdiv");

root.setThemes([
  am5themes_Animated.new(root)
]);

var value = 100;

function generateChartData() {
  var chartData = [];
  var firstDate = new Date();
  firstDate.setDate(firstDate.getDate() - 1000);
  firstDate.setHours(0, 0, 0, 0);

  for (var i = 0; i < 50; i++) {
    var newDate = new Date(firstDate);
    newDate.setSeconds(newDate.getSeconds() + i);

    value += (Math.random() < 0.5 ? 1 : -1) * Math.random() * 10;

    chartData.push({
      date: newDate.getTime(),
      value: value
    });
  }
  return chartData;
}

var data = generateChartData();

var chart = root.container.children.push(am5xy.XYChart.new(root, {
  focusable: true,
  panX: true,
  panY: true,
  wheelX: "panX",
  wheelY: "zoomX"
}));

var easing = am5.ease.linear;

var xAxis = chart.xAxes.push(am5xy.DateAxis.new(root, {
  maxDeviation: 0.5,
  groupData: false,
  extraMax:0.1,
  extraMin:-0.1,
  baseInterval: {
    timeUnit: "second",
    count: 1
  },
  renderer: am5xy.AxisRendererX.new(root, {
    minGridDistance: 50
  }),
  tooltip: am5.Tooltip.new(root, {})
}));

var yAxis = chart.yAxes.push(am5xy.ValueAxis.new(root, {
  renderer: am5xy.AxisRendererY.new(root, {})
}));

var series = chart.series.push(am5xy.LineSeries.new(root, {
  name: "Series 1",
  xAxis: xAxis,
  yAxis: yAxis,
  valueYField: "value",
  valueXField: "date",
  tooltip: am5.Tooltip.new(root, {
    pointerOrientation: "horizontal",
    labelText: "{valueY}"
  })
}));

data[data.length - 1].bullet = true;
series.data.setAll(data);

series.bullets.push(function(root, series, dataItem) {
  if (dataItem.dataContext.bullet) {
    var container = am5.Container.new(root, {});
    var circle0 = container.children.push(am5.Circle.new(root, {
      radius: 5,
      fill: am5.color(0xff0000)
    }));
    var circle1 = container.children.push(am5.Circle.new(root, {
      radius: 5,
      fill: am5.color(0xff0000)
    }));

    circle1.animate({
      key: "radius",
      to: 20,
      duration: 1000,
      easing: am5.ease.out(am5.ease.cubic),
      loops: Infinity
    });
    circle1.animate({
      key: "opacity",
      to: 0,
      from: 1,
      duration: 1000,
      easing: am5.ease.out(am5.ease.cubic),
      loops: Infinity
    });

    return am5.Bullet.new(root, {
      locationX:undefined,
      sprite: container
    })
  }
})

var cursor = chart.set("cursor", am5xy.XYCursor.new(root, {
  xAxis: xAxis
}));
cursor.lineY.set("visible", false);

var interval = setInterval(function () {
  addData();
}, 100)


function addData() {
  var lastDataItem = series.dataItems[series.dataItems.length - 1];

  var lastValue = lastDataItem.get("valueY");
  var newValue = value + ((Math.random() < 0.5 ? 1 : -1) * Math.random() * 5);
  var lastDate = new Date(lastDataItem.get("valueX"));
  var time = am5.time.add(new Date(lastDate), "second", 1).getTime();
  series.data.removeIndex(0);
  series.data.push({
    date: time,
    value: newValue
  })

  var newDataItem = series.dataItems[series.dataItems.length - 1];
  newDataItem.animate({
    key: "valueYWorking",
    to: newValue,
    from: lastValue,
    duration: 600,
    easing: easing
  });

  newDataItem.bullets = [];
  newDataItem.bullets[0] = lastDataItem.bullets[0];
  newDataItem.bullets[0].get("sprite").dataItem = newDataItem;
  lastDataItem.dataContext.bullet = false;
  lastDataItem.bullets = [];


  var animation = newDataItem.animate({
    key: "locationX",
    to: 0.5,
    from: -0.5,
    duration: 600
  });
  if (animation) {
    var tooltip = xAxis.get("tooltip");
    if (tooltip && !tooltip.isHidden()) {
      animation.events.on("stopped", function () {
        xAxis.updateTooltip();
      })
    }
  }
}

chart.appear(1000, 100);
```

**Comment by deicidemilan** [comment]
Sorry, this time I didn't test heap memory, I test Memory footprint in chrome Task Manager (shift+esc).
I did testing again, on your demo page (I have same result in my project). And heap memory is perfect, but Memory footprint still leaking if tab is focused.
I test few charts and settings, I have this problem now on this charts [Line Chart Adding Data Every Second](https://www.amcharts.com/demos/line-chart-adding-data-every-second/) and [Animated Bullet at the End of the Series
](https://www.amcharts.com/demos/animated-bullet-at-the-end-of-the-series/) They both have animated bullet, so I think it's something with animation. But I test it even without bullets `series.data.push(generateData())`
I don't know why, but Memory footprint is going up like crazy on this settings.

This is the things I do on page:
1) Open page.
2) Use zoom and pan 2-3 time (this is when memory start to grow up, and not clear).
3) Leave animation and tab focused (if not Memory footprint clear because tab not spend resources).

After 15 minutes it's growing up to 2gb+ memory.
I need tab to be focused all time. To update chart and follow state of data coming.
But application crashes in 10-15 minutes. Can you get idea why this happening?

This is print screen after 15 minutes on official demo page.

![This is an image](https://i.ibb.co/9GGqJdz/amcharts2.png)

this page have same problem (https://www.amcharts.com/demos/animated-bullet-at-the-end-of-the-series/)

**Comment by Pauan** [comment]
@deicidemilan Please test the code locally, not on the amCharts website.

We need to verify whether it is a problem with the V5 code or whether it's an unrelated problem with the amCharts website.

**Comment by deicidemilan** [comment]
@Pauan I test it locally all time, parallel with official because on fist spot, I was scare it's React.js problem.

My test was on:

1) Pure Html app
2) React app
3) Electron app
4) Official page
5) codepen.io

It's same results everywhere.

Did you get Memory footprint crash?

I use to test few chart lib, because I need for very serious app, and I like this chart most of all, but need to be active 24/7 without crashes :(

**Comment by martynasma** [comment]
@deicidemilan When crash occurs:

1) Do you have Dev Tools open in your browser? Does it happen if Dev Tools are closed?
2) Does this happen on multiple machines?
3) Do any other intensive processes take place on a machine you are testing with during tests?

**Comment by deicidemilan** [comment]
@martynasma

1) This is happening even if Dev Tools is closed. Even if is closed Task Manager and Dev Tools (nothing open).
2) Yes I tested on two machines, On first one tested on Linux and Windows 10, on second only Windows 10.
3) No Only this process, and system processes.

Machine I working on is:
Intel(R) Core(TM) i7-10700 CPU @ 2.90GHz
2x16GB DDR4 2933 MT/s Kingston
NVIDIA Quadro P1000
1 TB KINGSTON NVMe disk

Did you manage to catch this memory footprint leak. If not, do you want me to send video of how i get it?

**Comment by martynasma** [comment]
> Did you manage to catch this memory footprint leak. If not, do you want me to send video of how i get it?

We didn't.

Yes, please do send. Maybe that'll help us reproduce it.

**Comment by deicidemilan** [comment]
@martynasma

Here is a video for lib crashing on official demo page. It was the same on every app I tested, even on multiple systems (we tested on one more Windows 10)
I have cut the middle part of the video because it was too long. The application which uses your lib takes around 12 minutes to crash.

[Video](https://fonbgacrs-my.sharepoint.com/personal/aj20203047_student_fon_bg_ac_rs/_layouts/15/onedrive.aspx?id=%2Fpersonal%2Faj20203047%5Fstudent%5Ffon%5Fbg%5Fac%5Frs%2FDocuments%2FAmChart5%5FCrash2%2Ewmv&parent=%2Fpersonal%2Faj20203047%5Fstudent%5Ffon%5Fbg%5Fac%5Frs%2FDocuments)

**Comment by zeroin** [comment]
I doubt that the video will help, as we still have to reproduce it, but I can't view it anyway, I get some message like:

We're sorry, but [email] can't be found in the fonbgacrs-my.sharepoint.com directory. Please try again later, while we try to automatically fix this for you.

**Comment by deicidemilan** [comment]
@zeroin Sorry I will upload it again on Monday on youtube. This is a Microsoft cloud, but definitely have priority problem.
I will try to open to public now (it was yesterday). If not, in Monday from office I will upload it again.

**Comment by Pauan** [comment]
@deicidemilan We can sometimes reproduce the crash, but it often takes 30-60 minutes. Even then the Heap Snapshot shows that there is no memory leak in amCharts. So we suspect it's a bug in the browser, because it doesn't happen in Firefox.

**Comment by deicidemilan** [comment]
@Pauan @martynasma @zeroin

I didn't test in Firefox it's my mistake.
Now I tested even from home office, and on Edge, Opera, Electron and Firefox.
Turned out that chromium has problems. Chrome, Edge, Electron and Opera is on chromium engine. And all of them have memory footprint leak.
On Firefox I didn't get this problem. So probably chromium engine is the problem.

I got a crash on chromium related browsers in around 15 minutes. I uploaded a [video](https://www.youtube.com/watch?v=j694B4zUsuo&t=3s) again.

We have tested chart.js equally as amChart 5 on all platforms, and browsers.
Thing that caught my attention is that the chart.js library does not leak anywhere.

**Comment by martynasma** [comment]
@deicidemilan We have tracked the issue down to a [Chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1279394), which we reported and are waiting for a fix.

In the interim, we've released amCharts 5.0.18 which should fix the leak by applying a workaround.

We're not getting leaks/crashes anymore, but you might doublecheck if the upgrade fixes it in your setup, too.

I'm gonna go and close this issue, but feel free to reopen if you're still seeing the leak on your end.

**Comment by deicidemilan** [comment]
@martynasma I tested again charts on few environment, not detailed as before, but it's work perfect now on Linux with electron, and chrome browser. 

This is happening in amCharts 4 too. If you want to fix it.

We now wait Chromium to fix this.

If I found anything else, I will report to you.
